Avoid duplicated signatures when building transactions

grarco opened the pull request:
## Describe your changes

Closes #4203.

This PR modifies adds a new method on the `Tx` type to remove duplicated sections. This function is then called in the signing function before producing the wrapper signature (which is computed on all the sections) and the `dump_tx` function (in case the dumped tx needed to be signed offline).

To support this logic the `PartialEq` implementation for the `Authorization` type is now manually provided and ignores the way the signer is specified, focusing instead on the targets and the actual signature.

As a side note, also the `PartialEq` implementation for `SigningTxData` has been updated to account for the shielded hash.

## Checklist before merging 
- [ ] If this PR has some consensus breaking changes, I added the corresponding `breaking::` labels
    - This will require 2 reviewers to approve the changes
- [ ] If this PR requires changes to the docs or specs, a corresponding PR is opened in the `namada-docs` repo
    - Relevant PR if applies: 
- [ ] If this PR affects services such as `namada-indexer` or `namada-masp-indexer`, a corresponding PR is opened in that repo
    - Relevant PR if applies:

murisi commented on crates/tx/src/section.rs:
```diff
@@ -442,6 +441,39 @@ pub struct Authorization {
+            signatures: other_signatures,
+        } = other;
+
+        // Two authorizations are equal when they are computed over the same
```
This definition could work. But it has the implication that if `Authorization::signer` contains the variant `Signer::PubKeys` and someone tampers with one of the public keys in that list so that `Authorization::verify_signature` can no longer pass, this `PartialEq` implementation would still say that the tampered instance is still equal to the original. No? If so, is this desirable? Given that this function defines a partial equivalence relation, I guess `false` should be returned in those cases where there is insufficient information to check/conclude equality of the signers? No?

grarco replied:
That's correct. My reasoning was that once we establish the equality of the signature's targets and of the signatures themselves we can safely consider the two `Authorization`s to be the same even if the signers are expressed differently. In case of the signers being straight different (which we cause one of the two instances to fail the signature verification check) I believe we are looking at something else than just equality between the two instances: I guess one should check the integrity of the sections at the byte level using a digest or similar techniques but I don't have a super strong opinion on this (see also my answer to your next question since it's related to this one)

murisi commented on crates/tx/src/section.rs:
```diff
@@ -442,6 +441,39 @@ pub struct Authorization {
 }
 
+impl PartialEq for Authorization {
+    fn eq(&self, other: &Self) -> bool {
```
Does it matter that partially equivalent `Authorization`s might not have the same cryptographic hash? Or is this impossible? Could this sort of definition of partial equality confuse people in the future? (This definition seems to be calculated specifically to deal with our duplicate signatures issue (no?), but could be unintuitive in other cases/situations...)

murisi replied:
> I don't believe PartialEq should guarantee a byte-level match between the two instances but rather a logical equivalence (as I said in the previous question, when safety is critical one should explicitly rely on the digest of the expected instance).

Good point. My expectations of `PartialEq` were too restrictive.
> I could try to reintroduce the check on the signer with a more refined logic.

Thanks! This would be nice if it doesn't break the PR...
> My only concern is the case when a multisig signer is expressed as a set of keys in one instance and as an address in the other one: in this case there's no way I can resolve the equivalence without a query to storage (but there's a chance we always express multisig signers as a set of keys which would solve this issue).

Understood, I see the dilemma here. In this vein of thought, the comparison function could even be parametrized by a map from multisig addresses to public keys (given that the client must have known this in order to produce the signature sections in the first place). But this is probably overkill.

murisi reviewed (changes requested):
While the overall approach taken here to avoid duplicate signatures makes sense, checking whether `Authorization` sections are duplicates after they have been produced (by checking structural equality) seems less robust than not producing duplicates in the first place. But perhaps this (situation where a signing process is redundantly being done twice) is unavoidable...

grarco commented on crates/tx/src/section.rs:
```diff
@@ -442,6 +441,39 @@ pub struct Authorization {
+        // target(s) and the signatures match ,regardless of how the signer is
+        // expressed
+
+        // Check equivalence of the targets ignoring the specific ordering
```
Same as above, thanks!

grarco commented:
> While the overall approach taken here to avoid duplicate signatures makes sense, checking whether `Authorization` sections are duplicates after they have been produced (by checking structural equality) seems less robust than not producing duplicates in the first place. But perhaps this (situation where a signing process is redundantly being done twice) is unavoidable...

We already try to filter duplicated instanced of `SigningTxData` but there are cases where this logic fails (such as in case of different wrapper signers, even though we should only produce a single wrapper signature, see #3901, or the case of an offline procedure where we could get duplicated signatures from the members of a multisig account): hence the need to integrate this logic with a pruning mechanism for duplicated sections. I do agree with you though and we should try to improve the signing procedure over time to avoid producing duplicated signatures as much as possible

grarco commented:
@murisi I've pushed another commit to reintroduce the signer as part of the equivalence logic for `Authorization`. To do so I've implemented (yet again) a custom version of `PartialEq` for `Signer`, let me know if this addresses your concerns

murisi commented:
> > While the overall approach taken here to avoid duplicate signatures makes sense, checking whether `Authorization` sections are duplicates after they have been produced (by checking structural equality) seems less robust than not producing duplicates in the first place. But perhaps this (situation where a signing process is redundantly being done twice) is unavoidable...
> 
> We already try to filter duplicated instanced of `SigningTxData` but there are cases where this logic fails (such as in case of different wrapper signers, even though we should only produce a single wrapper signature, see #3901, or the case of an offline procedure where we could get duplicated signatures from the members of a multisig account): hence the need to integrate this logic with a pruning mechanism for duplicated sections. I do agree with you though and we should try to improve the signing procedure over time to avoid producing duplicated signatures as much as possible

Understood, I see your point: there are a lot of ways in which duplicate signatures can occur. And trying to catch duplicates during signing might actually be more complicated (or even more confusing) than removing them after signing.

murisi reviewed (approved):
Apologies for the delay. This PR looks good to me, thanks!

murisi commented:
> @murisi I've pushed another commit to reintroduce the signer as part of the equivalence logic for `Authorization`. To do so I've implemented (yet again) a custom version of `PartialEq` for `Signer`, let me know if this addresses your concerns

I have looked at the adjustments - they do address my concerns. Thanks!
